### ISSUE TITLE
Ability to better support odd scheduling time

### ISSUE THREAD

**Issue by Shadowsong27** [apache/airflow#10194]
**Description**

So I have been encountering odd time schedule request recently, say some updates need to be done
on every second and third Friday each month, 4pm.

Tried to search for solution but could not find a perfect ones, I can think of the following solutions:

1. use a cron and a separate DAG
2. embed trigger in the actual task logic, that is to check whether the code should be executed in that date, otherwise return success (equivalent to skip and mark success)

Drawbacks (at least the one i can think of):
1. it creates extra DAG, and when you have a bunch of these requests, it's not really clean. Besides, you need to add equal number of ExternalTaskSensor to preserve inter-DAG dependencies. It also however, will need to provision other dependent resources (such as Kafka consumers)

2. not very generic and reusable apparently, and if you are going for a full containerized solution - each Task is executed in container - you also need to start a container first to reach that simple if-else check, which seems a bit wasteful

**Use case / motivation**

Basically I have a bunch of extraction scripts that need to be executed at very different and very weird times, very straight forward (?), without the need to replicating provision tasks, too many inter-DAG dependencies

Potential solutions I have in mind:

- not touching Airflow core (workaround)

inherit and extend the Operators i am using - troublesome and needs to repeat for multiple operators

- touching Airflow core

an extra tuple params `allowed_dates` passed into the operators (BaseOperator), and then inject the date check right before the operator execution, essentially giving all operator the ability to be programmatically skipped in any DAG - might be an overkill

**Comment by JeffryMAC** [comment]
This has many requests (not all the same but the general idea is more scheduling abilities)
https://github.com/apache/airflow/issues/10123
https://github.com/apache/airflow/issues/9321
https://github.com/apache/airflow/issues/8649

and I'm adding my +1000 to having more scheduling ability.

I think at the end Airflow as workflow management system and as such the key functionality needed is flexible scheduling. In my opinion this should be one of the pillars of the tool.

**Comment by Shadowsong27** [comment]
I have just hacked some workaround since I really need this feature, it may be a bit raw but it's usable, using the first approach I mentioned above, with some abstractions. Basically I create a function that takes in any Airflow Operator, and returned a children of it, opening up an additional argument to skip dates. 

Code here:

```py
def flexify_operator(operator: Any) -> Any:

    class FlexibleOperator(operator):

        @apply_defaults
        def __init__(
                self,
                allowed_dates: Tuple = (),
                *args,
                **kwargs,
        ):
            super().__init__(*args, **kwargs)
            self.allowed_dates = allowed_dates

        def execute(self, context):
            execution_date = context.get('ds')
            if self.allowed_dates:

                if self.is_skipping_task(execution_date):
                    self.log.warning('Skipping the task since the date is not found in the allowed dates. ')
                    return

            # no allowed dates passed, or not skippingtreat as normal operator
            super().execute(context)

        def is_skipping_task(self, execution_date: str) -> bool:

            date_obj = dt.datetime.strptime(execution_date, '%Y-%m-%d')

            if date_obj.day not in self.allowed_dates:
                return True

            return False

    return FlexibleOperator

```

usage in DAG file:

```py
from airflow.operators.postgres_operator import PostgresOperator
PostgresOperator = utils.flexify_operator(PostgresOperator)

PostgresOperator(
    task_id=f"test",
    postgres_conn_id="conn_id",
    sql="SELECT * FROM public.test_demo",
    dag=dag,
    allowed_dates=(12, 13, 14)
)
```

So apparently this works only at day granularity, but it can be easily hacked to fulfill your needs, hope this helps.

**Comment by JeffryMAC** [comment]
I think you can utilise this to become a `FelixbileDateSensor` just like we have `WeekDaySensor`
@kaxil maybe you can give some pointers if that idea of a Sensor can be accepted as a core sensor?

**Comment by kaxil** [comment]
Have a look at https://lists.apache.org/thread.html/5bdf8415b78e65efb3219eb7f8e2afa37afa592341653eb053422055%40%3Cdev.airflow.apache.org%3E for some ideas from Max (the creator of Airflow)

and https://lists.apache.org/thread.html/re817fae1d9f05e9d3ca4c02ff49b0abd817859e3e3376208ac916c7f%40%3Cdev.airflow.apache.org%3E

**Comment by Shadowsong27** [comment]
@JeffryMAC By reading the threads above I feel my use case is closer to an operator based from `BaseBranchOperator` instead of a sensor approach, something like `FlexibleTimeBranchOperator` which allows you to branch DAG logic base on purely the execution datetime, by a customized func. We could have some pre-builds like `@monday` similar to `@daily` in DAG interval etc. 

I do agree that we should not introduce too much complexity to Airflow core, that's why I feel a plugin / operator like approach is probably better.

I provide a PR for this if you guys think this branch logic is useful.

**Comment by JeffryMAC** [comment]
@Shadowsong27 I don't know if it's the same case. having `@monday` isn't a "dynamic" it's something can be obtained with a single cron expression `0 0 * * MON`. The idea of better scheduling times is to allow to schedule jobs with intervals that can't be specified by a single cron expression.

**Comment by frosk1** [comment]
@Shadowsong27 it is nice to see more reminiscence and feedback on this topic. I have been gone the complete way from a outside enduser view to force airflow becoming more flexible in its scheduling mechanism: opened feature request[ #10449](https://github.com/apache/airflow/issues/10449), send draft via mail to dev@airflow.apache.org, etc. 

Within my feature request I point out the huge huge demand coming from endusers in terms of stackoverflow discussions, where  users not even see this being under development.

**Comment by Shadowsong27** [comment]
> @Shadowsong27 I don't know if it's the same case. having `@monday` isn't a "dynamic" it's something can be obtained with a single cron expression `0 0 * * MON`. The idea of better scheduling times is to allow to schedule jobs with intervals that can't be specified by a single cron expression.

Ok `@monday` is a bad example, guess pre-builts are not meaningful since if they are common enough to be a pre-built schedule, they most likely can be expressed in cron. 

I totally agreed on the rest, that's actually why I felt a custom function defined by users is the way to go here.

**Comment by Shadowsong27** [comment]
@frosk1 Yep I have been using Airflow since 2018, watched it grows to become the top choice in the market. This issue is not super blocking TBH, you can still kind of create one DAG per each odd schedule, but it creates a lot of pain, especially in most of the companys the DAG creation processes are still handled by data engineers (IMO, they should be done by the data consumers, via some sort of web UI).

That's why sometimes we will have to write our own plugins I guess, but I do see this as a potential Airflow core feature.

**Comment by ashb** [comment]
I've started a discussion thread on this on the dev mailing list to scope out what a solution to this will look like https://lists.apache.org/thread.html/rb4e004e68574e5fb77ee5b51f4fd5bfb4b3392d884c178bc767681bf%40%3Cdev.airflow.apache.org%3E

Use cases there would be ace (and feedback once we come up with a design)

**Comment by eladkal** [comment]
Please check Timetables
https://airflow.apache.org/docs/apache-airflow/stable/concepts/timetable.html
If a specific feature is missing you can open a new issue and explain
